lib/state_change_syncer.go

```diff
@@ -0,0 +1,726 @@
+	return EncoderTypeStateChangeEntry
+}
+
+// UnflushedStateSyncerBytes is used to keep track of the bytes that should be written to the state change file upon a db flush.
+type UnflushedStateSyncerBytes struct {
+	StateChangeBytes     []byte
+	StateChangeBytesSize uint64
+	// This is a list of the indexes of the state change bytes that should be written to the state change index file.
+	StateChangeOperationIndexes []uint64
+}
+
+// StateChangeSyncer is used to keep track of the state changes that should be written to the state change file.
+type StateChangeSyncer struct {
+	// The file that the state changes are written to.
+	StateChangeFile        *os.File
+	StateChangeMempoolFile *os.File
+	// The file that allows quick lookup of a StateChangeEntry given its index in the file.
+	// This is represented by a list of uint32s, where each uint32 is the offset of the state change entry in the state change file.
+	StateChangeIndexFile        *os.File
+	StateChangeFileSize         uint64
+	StateChangeMempoolIndexFile *os.File
+	StateChangeMempoolFileSize  uint64
+
+	// This map is used to keep track of the bytes should be written to the state change file upon a db flush.
+	// The ID of the flush is to track which entries should be written to the state change file upon flush completion.
+	// This is needed because many flushes can occur asynchronously during hypersync, and we need to make sure that
+	// we write the correct entries to the state change file.
+	// During blocksync, all flushes are synchronous, so we don't need to worry about this. As such, those flushes
+	// are given the uuid.Nil ID.
+	UnflushedBytes map[uuid.UUID]UnflushedStateSyncerBytes
+
+	// This map is used to keep track of the key and value pairs that state syncer has already tracked (and therefore
+	// don't need to be re-emitted to the state change file).
+	// The key is the stringifyed key of the entry, plus the operation type.
+	// The value is the badger entry that was flushed to the db.
+	MempoolKeyValueMap map[string][]byte
+	// This map tracks the keys that were flushed to the mempool in a single flush.
+	// This is used to determine if there are any tracked mempool transactions that have "disappeared" from the current
+	// mempool state. If so, state syncer should completely refresh the mempool.
+	MempoolFlushKeySet map[string]bool
+
+	// Tracks the flush IDs of the last block sync flush and the last mempool flush.
+	// These are not used during hypersync, as many flushes are being processed asynchronously.
+	BlockSyncFlushId uuid.UUID
+	MempoolFlushId   uuid.UUID
+
+	// Mutex to prevent concurrent writes to the state change file.
+	StateSyncerMutex *sync.Mutex
+	EntryCount       uint64
+
+	BlockHeight uint64
+
+	SyncType NodeSyncType
+
+	// During blocksync, we flush all entries by index to the state change file once the sync is complete.
+	// This is used to track whether this procedure has been initiated.
+	BlocksyncCompleteEntriesFlushed bool
```

Comment:
Explain the deeper reasons for why we do this optimization, rather than just flushing with each block. We want to give a lot of context for optimizations like these so that we can revisit them intelligently in the future. Otherwise, people are scared to change things like this for fear it will break things in unexpected ways.